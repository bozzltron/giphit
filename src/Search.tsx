import React, { useState, useEffect } from 'react';
import './App.css';
import { useSelector, useDispatch } from 'react-redux'
import { getGifs } from './redux/gifs/actions'
import { RootState } from './redux/reducer'
import useWebP from './useWebp'
import searchIcon from './search-icon.svg';
import Gif from './Gif'

function Search() {
  const supportsWebP = useWebP();
  const [query, setQuery] = useState('');
  const dispatch = useDispatch();
  const gifs = useSelector((state: RootState) => state.gifs)
  const ui = useSelector((state: RootState) => state.ui)
  const placeholders = [];
  for(let i=0; i<25; i++){
    placeholders.push("");
  }

  useEffect(() => {
    // Load trending gifs initially
    dispatch(getGifs('', 0))
  },[]);

  useEffect(() => {
    // Listen for scroll bottom to add more content
    let loadMoreGifs = () => {
      if ((window.innerHeight + window.scrollY) >= document.body.offsetHeight) {
        dispatch(getGifs(query, gifs.data.length));
      }
    }
    window.addEventListener('scroll', loadMoreGifs);
    return () => {
      window.removeEventListener('scroll', loadMoreGifs);
    }
  },[gifs]);

  return (
    <div className="app">
      <header className="search">
        <input type="search" placeholder="Search for gifs" onChange={(e)=>{ setQuery(e.target.value) }} />
        <button onClick={()=>{ dispatch(getGifs(query, 0)) }}><img src={searchIcon} alt="search icon" /></button>
      </header>
      <div className="container">
        {
          (gifs.data.length) ? gifs.data.map((gif: any, i:number)=>
            <Gif key={i} 
              id={gif.id}
              showOriginalId={ui.showOriginalId}
              preview={supportsWebP ? gif.images.preview_webp.url : gif.images.preview_gif.url} 
              original={supportsWebP ? gif.images.original.webp :  gif.images.original.url } 
              title={gif.title}  />
          ) : null
        }
        {
          gifs.isFetching || gifs.data.length === 0 ? placeholders.map((item, i)=><div key={i} className="item"><div className="placeholder"></div></div>) : null
        }
      </div>
    </div>
  );
}

export default Search;
