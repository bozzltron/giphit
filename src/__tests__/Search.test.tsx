import React from 'react';
import Search from '../Search';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'


it('renders correctly', () => {
  const initialState = {gifs: {isFetching: false, data: []}, ui: {showOriginalId:''}}
  const mockStore = configureStore()
  let store = mockStore(initialState);

  const tree = renderer
    .create(<Provider store={store}><Search /></Provider>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});