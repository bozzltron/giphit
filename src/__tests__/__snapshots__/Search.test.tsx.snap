// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="app"
>
  <header
    className="search"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        placeholder="Search for gifs"
        type="search"
      />
      <button>
        <img
          alt="search icon"
          src="search-icon.svg"
        />
      </button>
    </form>
  </header>
  <div
    className="container"
  >
    <div
      className="item"
    >
      <div
        className="placeholder"
      />
    </div>
    <div
      className="item"
    >
      <div
        className="placeholder"
      />
    </div>
    <div
      className="item"
    >
      <div
        className="placeholder"
      />
    </div>
    <div
      className="item"
    >
      <div
        className="placeholder"
      />
    </div>
    <div
      className="item"
    >
      <div
        className="placeholder"
      />
    </div>
    <div
      className="item"
    >
      <div
        className="placeholder"
      />
    </div>
    <div
      className="item"
    >
      <div
        className="placeholder"
      />
    </div>
    <div
      className="item"
    >
      <div
        className="placeholder"
      />
    </div>
    <div
      className="item"
    >
      <div
        className="placeholder"
      />
    </div>
    <div
      className="item"
    >
      <div
        className="placeholder"
      />
    </div>
    <div
      className="item"
    >
      <div
        className="placeholder"
      />
    </div>
    <div
      className="item"
    >
      <div
        className="placeholder"
      />
    </div>
    <div
      className="item"
    >
      <div
        className="placeholder"
      />
    </div>
    <div
      className="item"
    >
      <div
        className="placeholder"
      />
    </div>
    <div
      className="item"
    >
      <div
        className="placeholder"
      />
    </div>
    <div
      className="item"
    >
      <div
        className="placeholder"
      />
    </div>
    <div
      className="item"
    >
      <div
        className="placeholder"
      />
    </div>
    <div
      className="item"
    >
      <div
        className="placeholder"
      />
    </div>
    <div
      className="item"
    >
      <div
        className="placeholder"
      />
    </div>
    <div
      className="item"
    >
      <div
        className="placeholder"
      />
    </div>
    <div
      className="item"
    >
      <div
        className="placeholder"
      />
    </div>
    <div
      className="item"
    >
      <div
        className="placeholder"
      />
    </div>
    <div
      className="item"
    >
      <div
        className="placeholder"
      />
    </div>
    <div
      className="item"
    >
      <div
        className="placeholder"
      />
    </div>
    <div
      className="item"
    >
      <div
        className="placeholder"
      />
    </div>
  </div>
</div>
`;
